pkg/validations/upgradevalidations/immutableFields.go

```diff
@@ -169,3 +100,83 @@ func ValidateImmutableFields(ctx context.Context, k validations.KubectlClient, c
 
 	return provider.ValidateNewSpec(ctx, cluster, spec)
 }
+
+func ValidateGitOpsImmutableFields(ctx context.Context, k validations.KubectlClient, cluster *types.Cluster, clusterSpec *cluster.Spec, oldCluster *v1alpha1.Cluster) error {
+	if oldCluster.Spec.GitOpsRef == nil {
+		return nil
+	}
+
+	if !clusterSpec.Cluster.Spec.GitOpsRef.Equal(oldCluster.Spec.GitOpsRef) {
+		return errors.New("once cluster.spec.gitOpsRef is set, it is immutable")
+	}
+
+	switch clusterSpec.Cluster.Spec.GitOpsRef.Kind {
+	case v1alpha1.GitOpsConfigKind:
+		prevGitOps, err := k.GetEksaGitOpsConfig(ctx, clusterSpec.Cluster.Spec.GitOpsRef.Name, cluster.KubeconfigFile, clusterSpec.Cluster.Namespace)
+		if err != nil {
+			return err
+		}
+
+		if prevGitOps.Spec.Flux.Github.Owner != clusterSpec.GitOpsConfig.Spec.Flux.Github.Owner {
+			return errors.New("gitOps spec.flux.github.owner is immutable")
+		}
+		if prevGitOps.Spec.Flux.Github.Repository != clusterSpec.GitOpsConfig.Spec.Flux.Github.Repository {
+			return errors.New("gitOps spec.flux.github.repository is immutable")
+		}
+		if prevGitOps.Spec.Flux.Github.Personal != clusterSpec.GitOpsConfig.Spec.Flux.Github.Personal {
+			return errors.New("gitOps spec.flux.github.personal is immutable")
+		}
+		if clusterSpec.GitOpsConfig.Spec.Flux.Github.FluxSystemNamespace != "" && prevGitOps.Spec.Flux.Github.FluxSystemNamespace != clusterSpec.GitOpsConfig.Spec.Flux.Github.FluxSystemNamespace {
```

Comment:
what happens if the new cluster config `GitOpsConfig` spec includes a `""` `FluxSystemNamespace` (or `Branch`, or `ClusterConfigPath`) but the previous cluster spec `prevGitOps` has it set to a non-nil value? I believe that golang short-circuts logical AND so we'd only check if the new value is `""` and then never check if the old value was non-nil or equal, right? I realize this is just moving the old validations code and I very well might have written these validations :)

Reply:
talked to Joey -- will confirm that the defaults are set prior to the validation execution which would make this a moot point